opsdroid/tests/test_connector.py

```diff
@@ -0,0 +1,145 @@
+"""Test the opsdroid connector base class."""
+import pytest
+import asyncio
+
+import asynctest.mock as amock
+
+from opsdroid.core import OpsDroid
+from opsdroid.connector import Connector, register_event
+from opsdroid.events import Message, Reaction
+from opsdroid.cli.start import configure_lang
+
+configure_lang({})
+
+
+@pytest.fixture(scope="session")
+def loop():
+    return asyncio.new_event_loop()
+
+
+def test_init():
+    config = {"example_item": "test"}
+    connector = Connector(config, opsdroid=OpsDroid())
+    assert connector.default_target is None
+    assert connector.name == ""
+    assert connector.config["example_item"] == "test"
+
+
+def test_property():
+    opsdroid = amock.CoroutineMock()
+    connector = Connector({"name": "shell"}, opsdroid=opsdroid)
+    assert connector.configuration.get("name") == "shell"
+
+
+def test_connect(loop):
+    connector = Connector({}, opsdroid=OpsDroid())
+    with pytest.raises(NotImplementedError):
+        loop.run_until_complete(connector.connect())
+
+
+def test_listen(loop):
+    connector = Connector({}, opsdroid=OpsDroid())
+    with pytest.raises(NotImplementedError):
+        loop.run_until_complete(connector.listen())
+
+
+def test_respond(loop):
+    connector = Connector({}, opsdroid=OpsDroid())
+    with pytest.raises(TypeError):
+        loop.run_until_complete(connector.send(Message("")))
+
+
+def test_unsupported_event(loop):
+    connector = Connector({}, opsdroid=OpsDroid())
+    with pytest.raises(TypeError):
+        loop.run_until_complete(connector.send(Reaction("emoji")))
+
+
+def test_incorrect_event():
+    class NotanEvent:
+        pass
+
+    class MyConnector(Connector):
+        @register_event(NotanEvent)
+        def send_my_event(self, event):
+            pass
+
+    with pytest.raises(TypeError):
+        MyConnector()
+
+
+def test_event_subclasses():
+    class MyEvent(Message):
+        pass
+
+    class MyConnector(Connector):
+        @register_event(Message, include_subclasses=True)
+        def send_my_event(self, event):
+            pass
+
+    c = MyConnector({})
+    assert MyEvent in c.events
+
+
+"""Test the async methods of the opsdroid connector base class."""
```

Comment:
Why is the docstring is in the middle of the file?

Reply:
sorry, this is WIP. I wanted to create a draft PR first.

Reply:
then I put the functions back in classes, that way the docstrings make sense

Reply:
IMHO the classes are needed only for unittest tests style. Functions are more pytest way

Reply:
okay, but the classes may refer to classes we test. I can remove the classes, but then do what with the docstrings which indicate what we are testing in the next few tests.

Reply:
I know that functions are more pytest way, but I think we could have classes to do the tests. If I'm not mistaken we have merged some pytest tests within a class 👍

Reply:
then I leave in the classes. I think they can be useful for interpreting what we are testing.